tsl/src/remote/dist_copy.c

```diff
@@ -1248,355 +1223,388 @@ calculate_hyperspace_point_from_binary(Datum *values, bool *nulls, const Hypersp
+				 * CopyData message and flush the data. Do it on all
+				 * connections, even though only one connection might be
+				 * affected. */
+				write_copy_data_end(context);
+				/* No need to exit out of the COPY_IN mode in order to create
+				 * the chunk on the same connection; it is handled
+				 * automatically by remote_dist_txn_get_connection(). */
+				chunk = ts_hypertable_create_chunk_for_point(ht, p, &found);
 			}
+			else
+				found = true;
 
 			/*
-			 * Send out the current batch.
+			 * Get the filtered list of "available" DNs for this chunk but only if it's replicated.
+			 * We only fetch the filtered list once. Assuming that inserts will typically go to the
+			 * same chunk we should be able to reuse this filtered list a few more times
+			 *
+			 * The worse case scenario is one in which INSERT1 goes into CHUNK1, INSERT2 goes into
+			 * CHUNK2, INSERT3 goes into CHUNK1,... in which case we will end up refreshing the list
+			 * everytime
+			 *
+			 * We will also enter the below loop if we KNOW that any of the DNs has been marked
+			 * unavailable before we started this transaction. If not, then we know that every
+			 * chunk's datanode list is fine and no stale chunk metadata updates are needed.
 			 */
-			remote_copy_process_and_send_data(context);
+			if (context->dns_unavailable && found && ht->fd.replication_factor > 1)
+			{
+				MemoryContext oldmctx = MemoryContextSwitchTo(context->mctx);
 
-			processed += context->batch_row_count;
-			context->batch_row_count = 0;
-			context->batch_size_bytes = 0;
-			context->batch_ordinal++;
-			MemoryContextReset(batch_context);
+				if (chunk_id != chunk->fd.id)
+					chunk_id = INVALID_CHUNK_ID;
 
-			if (eof)
-			{
-				break;
+				if (chunk_id == INVALID_CHUNK_ID)
+				{
+					if (chunk_data_nodes)
+						list_free(chunk_data_nodes);
+					chunk_data_nodes =
+						ts_chunk_data_node_scan_by_chunk_id_filter(chunk->fd.id,
+																   CurrentMemoryContext);
+					chunk_id = chunk->fd.id;
+				}
+
+				Assert(chunk_id == chunk->fd.id);
+				Assert(chunk_data_nodes != NIL);
+				/*
+				 * If the chunk was not created as part of this insert, we need to check whether any
+				 * of the chunk's data nodes are currently unavailable and in that case consider the
+				 * chunk stale on those data nodes. Do that by removing the AN's chunk-datanode
+				 * mapping for the unavailable data nodes.
+				 *
+				 * Note that the metadata will only get updated once since we assign the chunk's
+				 * data_node list to the list of available DNs the first time this
+				 * dist_update_stale_chunk_metadata API gets called. So both chunk_data_nodes and
+				 * chunk->data_nodes will point to the same list and no subsequent metadata updates
+				 * will occur.
+				 */
+				if (ht->fd.replication_factor > list_length(chunk_data_nodes))
+					ts_cm_functions->dist_update_stale_chunk_metadata(chunk, chunk_data_nodes);
```

Comment:
This is not changed by your PR, but can we update metadata once per batch instead of once per row?

Reply:
I think @nikkhils already did some changes here, although I am not sure I fully follow what changed. 

Perhaps better to do those changes separately, with @nikkhils's input.

Reply:
@akuzm @erimatnor this metadata update change is already in master and 2.10.x branch